homeassistant/components/here_travel_time/__init__.py

```diff
@@ -1 +1,165 @@
-"""The here_travel_time component."""
+"""The HERE Travel Time integration."""
+from __future__ import annotations
+
+from datetime import datetime, time, timedelta
+import logging
+
+import async_timeout
+from herepy import NoRouteFoundError, RouteMode, RoutingApi, RoutingResponse
+
+from homeassistant.const import ATTR_ATTRIBUTION, CONF_UNIT_SYSTEM_IMPERIAL, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.location import find_coordinates
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+from homeassistant.util import dt
+
+from .const import (
+    ATTR_DESTINATION,
+    ATTR_DESTINATION_NAME,
+    ATTR_DISTANCE,
+    ATTR_DURATION,
+    ATTR_DURATION_IN_TRAFFIC,
+    ATTR_ORIGIN,
+    ATTR_ORIGIN_NAME,
+    ATTR_ROUTE,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    NO_ROUTE_ERROR_MESSAGE,
+    TRAFFIC_MODE_ENABLED,
+    TRAVEL_MODES_VEHICLE,
+)
+from .model import HERERoutingData, HERETravelTimeConfig
+
+PLATFORMS = [Platform.SENSOR]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class HereTravelTimeDataUpdateCoordinator(DataUpdateCoordinator):
+    """HERETravelTime DataUpdateCoordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        api: RoutingApi,
+        config: HERETravelTimeConfig,
+    ) -> None:
+        """Initialize."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=timedelta(seconds=DEFAULT_SCAN_INTERVAL),
+        )
+        self._api = api
+        self.config = config
+
+    async def _async_update_data(self) -> HERERoutingData | None:
+        """Get the latest data from the HERE Routing API."""
+        try:
+            async with async_timeout.timeout(10):
+                return await self.hass.async_add_executor_job(self._update)
+        except NoRouteFoundError as error:
+            raise UpdateFailed(NO_ROUTE_ERROR_MESSAGE) from error
+
+    def _update(self) -> HERERoutingData | None:
+        """Get the latest data from the HERE Routing API."""
+        if self.config.origin_entity_id is not None:
+            origin = find_coordinates(self.hass, self.config.origin_entity_id)
+        else:
+            origin = self.config.origin
+
+        if self.config.destination_entity_id is not None:
+            destination = find_coordinates(self.hass, self.config.destination_entity_id)
+        else:
+            destination = self.config.destination
+        if destination is not None and origin is not None:
+            here_formatted_destination = destination.split(",")
+            here_formatted_origin = origin.split(",")
+            arrival: str | None = None
+            departure: str | None = "now"
```

Comment:
Also I think the typing can get simpler given it can't be `None`:

```suggestion
            departure = "now"
```

Reply:
Actually the code is currently broken.

For `arrival` and `departure` is is possible to define either `None` or a timestamp. If it is a timestamp it has to be parsed into the format the HERE api expects.

It is possible that either `arrival` or `destination` is `None`.

If both are `None` than departure has to be set to `now` which is a somewhat hidden specification of the HERE api.

Reply:
So I think the code has to be reverted to:

```python
arrival: str | None = None
departure: str | None = None
if self.config.arrival is not None:
    arrival = convert_time_to_isodate(self.config.arrival)
if self.config.departure is not None:
    departure = convert_time_to_isodate(self.config.departure)

if arrival is None and departure is None:
    departure = "now"
```

Reply:
Got it, makes sense.